pkg/portutil/port_allocate_linux.go

```diff
@@ -39,27 +39,60 @@ func filter(ss []procnet.NetworkDetail, filterFunc func(detail procnet.NetworkDe
 }
 
 func portAllocate(protocol string, ip string, count uint64) (uint64, uint64, error) {
-	netprocData, err := procnet.ReadStatsFileData(protocol)
+	usedPort, err := getUsedPorts(ip, protocol)
+
 	if err != nil {
 		return 0, 0, err
 	}
-	netprocItems := procnet.Parse(netprocData)
+
+	start := uint64(allocateStart)
+	if count > uint64(allocateEnd-allocateStart+1) {
+		return 0, 0, fmt.Errorf("can not allocate %d ports", count)
+	}
+	for start < allocateEnd {
+		needReturn := true
+		for i := start; i < start+count; i++ {
+			if _, ok := usedPort[i]; ok {
+				needReturn = false
+				break
+			}
+		}
+		if needReturn {
+			return start, start + count - 1, nil
+		}
+		start += count
+	}
+	return 0, 0, fmt.Errorf("there is not enough %d free ports", count)
+}
+
+func getUsedPorts(ip string, protocol string) (map[uint64]bool, error) {
+	netprocItems := []procnet.NetworkDetail{}
+
+	if protocol == "tcp" || protocol == "udp" {
+		netprocData, err := procnet.ReadStatsFileData(protocol)
+		if err != nil {
+			return nil, err
+		}
+		netprocItems = append(netprocItems, procnet.Parse(netprocData)...)
+	}
```

Comment:
It appears that the SCTP protocol is currently unsupported, therefore no further processing is required.

Reply:
I could be wrong but from this line https://github.com/containerd/nerdctl/blob/main/pkg/portutil/portutil.go#L61, it seems like users could use `sctp`. There is also a test to check the parsing of `sctp` https://github.com/containerd/nerdctl/blob/main/pkg/portutil/portutil_test.go#L348.

Reply:
I only checkd in https://github.com/containerd/nerdctl/blob/main/pkg/portutil/procnet/procnet_linux.go#L38 , Let me check it.

Reply:
I carefully checked the code and did not find an implementation that supports the SCTP protocol. The [line](https://github.com/containerd/nerdctl/blob/main/pkg/portutil/portutil.go#L61) just supports parsing command line arguments.

Reply:
> I carefully checked the code and did not find an implementation that supports the SCTP protocol. The [line](https://github.com/containerd/nerdctl/blob/main/pkg/portutil/portutil.go#L61) just supports parsing command line arguments.

Additionally there a unit test that checks the parsing of "sctp" https://github.com/containerd/nerdctl/blob/main/pkg/portutil/portutil_test.go#L348, so without the protocol check below, the [unit test fails](https://github.com/containerd/nerdctl/actions/runs/4811153181/jobs/8564771421#step:4:463). 
```	
        if protocol == "tcp" || protocol == "udp" {
		netprocData, err := procnet.ReadStatsFileData(protocol)
		if err != nil {
			return nil, err
		}
		netprocItems = append(netprocItems, procnet.Parse(netprocData)...)
	}
```

As a side note, if sctp is not supported then why do we support parsing it in the [ParseFlagP](https://github.com/containerd/nerdctl/blob/main/pkg/portutil/portutil.go#L52) ?

Reply:
Yep..., let it works

Reply:
I quickly checked, it looks like `SCTP` is not supported but it is but it is not blocking either. IMHO, we have to add a comment [here](https://github.com/containerd/nerdctl/pull/2190/files#diff-b2a4351f70be5b1bee5c564f5ddcff9b98cdfec8f24ab1a4dd6a6df30bceb515R61) to highlight this point  ( It's misleading without a comment)

Reply:
Updated.